src/plugins.ts

```diff
@@ -0,0 +1,95 @@
+import { SchemaBuilder, NexusAcceptedTypeDef } from "./builder";
+import { id } from "./utils";
+
+/**
+ * This is a read-only builder-like api exposed to plugins in the onInstall
+ * hook. It proxies the Nexus Builder which is a much bigger beast that we do
+ * not want to directly expose to plugins.
+ */
+type BuilderFacade = {
+  hasType: (typeName: string) => boolean;
+};
+
+/**
+ * This is the Neuxs Plugin interface that allows users to extend Nexus at
+ * particular extension points. Plugins are just functions that receive hooks
+ * which can then be registered upon with callbacks.
+ */
+export type Plugin = (hooks: Hooks) => void;
+
+type Hooks = {
+  onInstall: OnInstallHook;
+};
+
+type OnInstallHook = (onInstallHandler: OnInstallHandler) => void;
+
+type OnInstallHandler = (
+  builder: BuilderFacade
+) => {
+  types: NexusAcceptedTypeDef[];
+};
```

Comment:
I intentionally made this 1) extensible and 2) strict.

By returning a record we allow for `onInstall` to do more things in the future. We can deprecate `types` key in favour of new ones that separate concepts, or whatever. Who knows.

Currently the only reason to use `onInstall` hook is to contribute type definitions. Rather than taking an optional approach and thus opening user up to error (e.g. accidental non-return) we just enforce it. We can easily change our minds later, relaxing the requirement, because turning this into a nullable would be backwards compatible.